suprb/rule/matching.py

```diff
@@ -78,3 +79,162 @@ def min_range(self, min_range: float):
+        diff = np.maximum(diff, -diff)
+        return np.prod(diff)
+
+    def copy(self):
+        return UnorderedBound(self.bounds.copy())
+
+    def _validate_bounds(self, X: np.ndarray):
+        """Validates that bounds have the correct shape."""
+
+        if self.bounds.shape[1] != 2:
+            raise ValueError(f"specified bounds are not of shape (-1, 2), but {self.bounds.shape}")
+
+        if self.bounds.shape[0] != X.shape[1]:
+            raise ValueError(f"bounds- and input data dimension mismatch: {self.bounds.shape[0]} != {X.shape[1]}")
+
+    def clip(self, bounds: np.ndarray):
+        self.bounds.clip(-1, 1)
+
+    def min_range(self, min_range: float):
+        diff = self.bounds[:, 1] - self.bounds[:, 0]
+        if min_range > 0:
+            invalid_indices = np.argwhere((diff < min_range) & (-diff < min_range))
+            # Tuple of shape (lower, upper)
+            invalid_indices_l = np.argwhere((diff[invalid_indices] > -diff[invalid_indices]))
+            # Tuple of shape (upper, lower)
+            invalid_indices_r = np.argwhere((diff[invalid_indices] <= -diff[invalid_indices]))
+
+            # Increase Range for tuples of shape (lower, upper)
+            self.bounds[invalid_indices_l, 0] += min_range / 2
+            self.bounds[invalid_indices_l, 1] -= min_range / 2
+
+            # Increase Range for tuples of shape (upper, lower)
+            self.bounds[invalid_indices_r, 0] -= min_range / 2
+            self.bounds[invalid_indices_r, 1] += min_range / 2
+
+
+class CentreSpread(MatchingFunction):
+    """
+    A standard interval-based matching function producing multi-dimensional
+    hyperrectangular conditions. In effect, a centre (c) and
+    a spread (s) are specified for each dimension.
+    An example x is matched iff c_i - s_i <= x_i <= c_i + s_i for all dimensions i
+    """
+
+    def __init__(self, bounds: np.ndarray):
+        self.bounds = bounds
+
+    def __call__(self, X: np.ndarray):
+        return np.all(((self.bounds[:, 0] - self.bounds[:, 1]) <= X) &
+                      (X <= (self.bounds[:, 0] + self.bounds[:, 1])), axis=1)
+
+    @property
+    def volume_(self):
+        """Calculates the volume of the interval."""
+        lower = self.bounds[:, 0] - self.bounds[:, 1]
+        higher = self.bounds[:, 0] + self.bounds[:, 1]
+        lower = lower.clip(-1, 1)
+        higher = higher.clip(-1, 1)
+        diff = higher - lower
+        return np.prod(diff)
+
+    def copy(self):
+        return CentreSpread(self.bounds.copy())
+
+    def _validate_bounds(self, X: np.ndarray):
+        """Validates that bounds have the correct shape."""
+
+        if self.bounds.shape[1] != 2:
+            raise ValueError(f"specified bounds are not of shape (-1, 2), but {self.bounds.shape}")
+
+        if self.bounds.shape[0] != X.shape[1]:
+            raise ValueError(f"bounds- and input data dimension mismatch: {self.bounds.shape[0]} != {X.shape[1]}")
```

Comment:
If we use this in multiple classes that inherit from `MatchingFunction` we can move this into the BaseClass itself and remove it from here